### PR TITLE
[SPARK-37258][K8S][BUILD] Upgrade kubernetes-client to 5.12.0

### PR DESCRIPTION
**Pull request by Yikun**
### What changes were proposed in this pull request?
This patch aims to upgrade `kubernetes-client` from 5.10.2 to 5.12.0

Hightlight changes:
- v5.12.0: Update Fabric8 Kubernetes Model to v1.23.0, bump jackson-datatype-jsr310 to [2.13.1](https://github.com/fabric8io/kubernetes-client/commit/43562675d36a7240825453370f5d540370e84cbb).
- v5.11.0: Introduce the Volcano extension, it would be useful for users who want to use Volcano as customized scheduler in Spark on K8S.
- v5.11.0: Breaking change: an abstraction layer added over okHttp

There are also some changes to make sure it is compatiable to create http client with custom dispatcher according to https://github.com/fabric8io/kubernetes-client/issues/3663#issuecomment-997214403 suggestion.

### Why are the changes needed?

This will bring several bug fixes and improvements (such as volcano support, Fabric8 Kuberentes model 1.23 support), see more in:
- https://github.com/fabric8io/kubernetes-client/releases/tag/v5.11.0
- https://github.com/fabric8io/kubernetes-client/releases/tag/v5.11.1
- https://github.com/fabric8io/kubernetes-client/releases/tag/v5.11.2
- https://github.com/fabric8io/kubernetes-client/releases/tag/v5.12.0

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Pass the CIs.
- Running a spark-pi job to validate manually.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Is it for Volcano?
> Fix #3579: Add support for Volcano extension

**Comment by Yikun** [comment]
There are some breaking changes in the API of k8s-cli v5.11, looks like it breaks to **create httpclient with custom dispatcher** [like what we did here](https://github.com/apache/spark/blob/19227983e91a54b5f27ade6412dad1088dfcba9e/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala#L110) .

So we need to do some changes to make it work.

issue: https://github.com/fabric8io/kubernetes-client/issues/3663

**Comment by Yikun** [comment]
> Is it for Volcano?

@dongjoon-hyun Yes~The volcano extesnion was introduced in k8s-cli v5.11.

**Comment by Yikun** [comment]
Retest this please

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala
Comment (Yikun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/SparkKubernetesClientFactory.scala:72]: Here are some backgroud on this note (also add the jira: [SPARK-37687](https://issues.apache.org/jira/browse/SPARK-37687)):

- There are [some problems (such as IPV6 based cluster support)](https://github.com/fabric8io/kubernetes-client/issues/2632) on okhttpclient v3, but it's a little bit diffcult to upgrade to v4 [1]
- Kubernetes client are also consider to[ support other clients](https://github.com/fabric8io/kubernetes-client/issues/3663#issuecomment-997402993) rather than single okhttpclient.
- Kubernetes client[ add a abstract layer](https://github.com/fabric8io/kubernetes-client/issues/3547) to support variety httpclient, okhttp client as one of supported http clients.

So, we better to consider to cleanup okhttpclient direct usage and use the httpclient which kubernetes client diret supported to reduce the potential risk in future upgrade.

**Comment by Yikun** [comment]
cc @dongjoon-hyun @holdenk @attilapiros

**Comment by Yikun** [comment]
cc @dongjoon-hyun @holdenk @attilapiros

**Comment by Yikun** [comment]
retest this please

**Review by dongjoon-hyun — COMMENTED** [review]
Actually, I'm a little reluctant due to the breaking change. Shall we hold on and see the next release?

**Comment by Yikun** [comment]
@dongjoon-hyun I can understand your concerns for sure, it makes sense to wait for next release to avoid pontential flaky factor of breaking changes.

But there is one thing I need to clarify, even in the next version we will still face these changes to make okhttpclient work. BTW, these changes are changed according to https://github.com/fabric8io/kubernetes-client/issues/3663#issuecomment-997214403 .

**Comment by klaus-xiong** [comment]
@Yikun Do you have a time line for the 'Cleanup direct usage of OkHttpClient' ?

**Comment by Yikun** [comment]
@klaus-xiong I think the deadline depends on when the new other httpclient to replace okhttpclient are supported in k8s client, and also this replacement meet the requirement of Spark on K8S.

It bring some potential issue when k8s-cli to stop okhttpclient support (as https://github.com/fabric8io/kubernetes-client/issues/3547 mentioned, OkHttpClient has some known issue and it's not easy to upgrade OkHttpClient), it may not be a good choice if we still keep the direct OkHttpClient usage in Spark.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Thank you for your patience. Could you update this PR with 5.12.0? After that, I believe it's time to merge.

- https://github.com/fabric8io/kubernetes-client/releases/tag/v5.12.0
```
Update Fabric8 Kubernetes Model to v1.23.0
```

**Comment by Yikun** [comment]
@dongjoon-hyun Sure, will upgrade soon. Thanks!

**Comment by dongjoon-hyun** [comment]
Thanks!

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @Yikun .
Merged to master for Apache Spark 3.3.

**Pull request closed without merging**
